File: clients/kotlin/src/test/kotlin/SchedulerTest.kt
import dev.vexilla.*
import kotlinx.datetime.*
import org.junit.jupiter.api.Test
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.days
import kotlin.test.*


//@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class SchedulerTest {

    @Test
    fun `should be active with an empty schedule`() {
        val schedule = Schedule(
            start = 0,
            end = 0,
            timezone = "UTC",
            timeType = ScheduleTimeType.NONE,
            startTime = 0,
            endTime = 0
        )
        assertTrue(Scheduler.isScheduleActive(schedule, ScheduleType.EMPTY))
    }


    @Test
    fun `should handle start_end single day schedules`() {
        val now = Clock.System.now()
        val zeroDay = Instant.fromEpochMilliseconds(0)

        for (hour in 0..23) {

            val mockedNow =
                LocalDateTime(
                    now.toLocalDateTime(TimeZone.UTC).date,
                    LocalTime(hour, 0, 0, 0)
                ).toInstant(
                    TimeZone.UTC
                )

            val zeroDayWithMockedTime = Instant.fromEpochMilliseconds(0).plus(hour.hours)

            val beforeSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeSchedule, ScheduleType.GLOBAL, mockedNow)

            assertFalse(isBeforeScheduleActive, "Hour: ${hour}: Before test failed")

            val duringSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds()
            )

            val isDuringScheduleActive =
                Scheduler.isScheduleActiveWithNow(duringSchedule, ScheduleType.GLOBAL, mockedNow)

            assertTrue(isDuringScheduleActive, "Hour: ${hour}: During test failed")


            val afterSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterScheduleActive = Scheduler.isScheduleActiveWithNow(afterSchedule, ScheduleType.GLOBAL, mockedNow)

            assertFalse(isAfterScheduleActive, "Hour: ${hour}: During test failed")
        }
    }

    @Test
    fun `should handle start_end multi day schedules`() {
        val now = Clock.System.now()
        val zeroDay = Instant.fromEpochMilliseconds(0)

        for (hour in 0..23) {

            val mockedNow =
                LocalDateTime(
                    now.toLocalDateTime(TimeZone.UTC).date,
                    LocalTime(hour, 0, 0, 0)
                ).toInstant(
                    TimeZone.UTC
                )

            val zeroDayWithMockedTime = Instant.fromEpochMilliseconds(0).plus(hour.hours)

            val beforeSchedule = Schedule(
                start = mockedNow.plus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(3.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeSchedule, ScheduleType.GLOBAL, mockedNow)

            assertFalse(isBeforeScheduleActive)

            val duringSchedule = Schedule(
                start = mockedNow.minus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds()
            )

            val isDuringScheduleActive =
                Scheduler.isScheduleActiveWithNow(duringSchedule, ScheduleType.GLOBAL, mockedNow)

            assertTrue(isDuringScheduleActive)


            val afterSchedule = Schedule(
                start = mockedNow.minus(3.days).toEpochMilliseconds(),
                end = mockedNow.minus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.START_END,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterScheduleActive = Scheduler.isScheduleActiveWithNow(afterSchedule, ScheduleType.GLOBAL, mockedNow)

            assertFalse(isAfterScheduleActive)
        }
    }


    @Test
    fun `should handle daily single day schedules`() {
        for (hour in 0..23) {
            val now = Clock.System.now().toLocalDateTime(TimeZone.UTC)

            val mockedNow =
                LocalDateTime(
                    now.date,
                    LocalTime(hour, 0)
                ).toInstant(
                    TimeZone.UTC
                )

            val zeroDayWithMockedTime = Instant.fromEpochMilliseconds(0).plus(hour.hours)

            val beforeWholeSchedule = Schedule(
                start = mockedNow.plus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeWholeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeWholeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isBeforeWholeScheduleActive, "Hour: ${hour}: Before whole test failed")

            val beforeSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isBeforeScheduleActive, "Hour: ${hour}: Before test failed")

            val duringSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds()
            )

            val isDuringScheduleActive =
                Scheduler.isScheduleActiveWithNow(duringSchedule, ScheduleType.GLOBAL, mockedNow)
            assertTrue(isDuringScheduleActive, "Hour: ${hour}: During test failed")

            val afterSchedule = Schedule(
                start = mockedNow.toEpochMilliseconds(),
                end = mockedNow.toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterScheduleActive = Scheduler.isScheduleActiveWithNow(afterSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isAfterScheduleActive, "Hour: ${hour}: After test failed")

            val afterWholeSchedule = Schedule(
                start = mockedNow.minus(1.days).toEpochMilliseconds(),
                end = mockedNow.minus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterWholeScheduleActive =
                Scheduler.isScheduleActiveWithNow(afterWholeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isAfterWholeScheduleActive, "Hour: ${hour}: After whole test failed")
        }
    }


    @Test
    fun `should handle daily multi day schedules`() {
        for (hour in 0..23) {
            val now = Clock.System.now().toLocalDateTime(TimeZone.UTC)

            val mockedNow =
                LocalDateTime(
                    now.date,
                    LocalTime(hour, 0)
                ).toInstant(
                    TimeZone.UTC
                )

            val zeroDayWithMockedTime = Instant.fromEpochMilliseconds(0).plus(hour.hours)

            val beforeWholeSchedule = Schedule(
                start = mockedNow.plus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(3.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeWholeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeWholeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isBeforeWholeScheduleActive, "Hour: ${hour}: Before whole test failed")

            val beforeSchedule = Schedule(
                start = mockedNow.minus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(3.hours).toEpochMilliseconds()
            )

            val isBeforeScheduleActive =
                Scheduler.isScheduleActiveWithNow(beforeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isBeforeScheduleActive, "Hour: ${hour}: Before test failed")

            val duringSchedule = Schedule(
                start = mockedNow.minus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.plus(1.hours).toEpochMilliseconds()
            )

            val isDuringScheduleActive =
                Scheduler.isScheduleActiveWithNow(duringSchedule, ScheduleType.GLOBAL, mockedNow)
            assertTrue(isDuringScheduleActive, "Hour: ${hour}: During test failed")

            val afterSchedule = Schedule(
                start = mockedNow.minus(1.days).toEpochMilliseconds(),
                end = mockedNow.plus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterScheduleActive = Scheduler.isScheduleActiveWithNow(afterSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isAfterScheduleActive, "Hour: ${hour}: After test failed")

            val afterWholeSchedule = Schedule(
                start = mockedNow.minus(3.days).toEpochMilliseconds(),
                end = mockedNow.minus(1.days).toEpochMilliseconds(),
                timezone = "UTC",
                timeType = ScheduleTimeType.DAILY,
                startTime = zeroDayWithMockedTime.minus(3.hours).toEpochMilliseconds(),
                endTime = zeroDayWithMockedTime.minus(1.hours).toEpochMilliseconds()
            )

            val isAfterWholeScheduleActive =
                Scheduler.isScheduleActiveWithNow(afterWholeSchedule, ScheduleType.GLOBAL, mockedNow)
            assertFalse(isAfterWholeScheduleActive, "Hour: ${hour}: After whole test failed")
        }
    }
}

File: clients/elixir/test/scheduler_test.exs
defmodule SchedulerTest do
  require ScheduleType
  require TimeType
  use ExUnit.Case
  doctest Scheduler


  test "handles empty schedule" do
    empty_schedule = %Schedule{
      start: 0,
      end: 0,
      timezone: "UTC",
      timeType: "none",
      startTime: 0,
      endTime: 0,
    } |> Jason.encode!() |> Jason.decode!()

    should_empty = Scheduler.is_schedule_active(empty_schedule, ScheduleType.empty(), Timex.now())

    assert should_empty == true
  end

  test "handles empty time schedules" do

    now = Timex.now()

    before_schedule = %Schedule{
      start: Timex.add(now, Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
      end: Timex.add(now, Timex.Duration.from_days(3)) |> Timex.to_unix |> Kernel.*(1000),
      timezone: "UTC",
      timeType: TimeType.none(),
      startTime: 0,
      endTime: 0,
    } |> Jason.encode!() |> Jason.decode!()

    during_schedule = %Schedule{
      start: Timex.add(now, Timex.Duration.from_days(-1)) |> Timex.to_unix |> Kernel.*(1000),
      end: Timex.add(now, Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
      timezone: "UTC",
      timeType: TimeType.none(),
      startTime: 0,
      endTime: 0,
    } |> Jason.encode!() |> Jason.decode!()

    after_schedule = %Schedule{
      start: Timex.add(now, Timex.Duration.from_days(-3)) |> Timex.to_unix |> Kernel.*(1000),
      end: Timex.add(now, Timex.Duration.from_days(-1)) |> Timex.to_unix |> Kernel.*(1000),
      timezone: "UTC",
      timeType: TimeType.none(),
      startTime: 0,
      endTime: 0,
    } |> Jason.encode!() |> Jason.decode!()

    should_before = Scheduler.is_schedule_active(before_schedule, ScheduleType.global(), now)
    assert should_before == false

    should_during = Scheduler.is_schedule_active(during_schedule, ScheduleType.global(), now)
    assert should_during == true

    should_after = Scheduler.is_schedule_active(after_schedule, ScheduleType.global(), now)
    assert should_after == false
  end

  test "handles start/end single day schedules" do

    zero_day = Timex.from_unix(0)

    Enum.each(0..23, fn(hour) ->

      mocked_now = Timex.now()
      |> Timex.set(
        hour: hour,
        minute: 0,
        second: 0,
        microsecond: 0
      )

      zero_day_with_mocked_time = zero_day |> Timex.set(
        hour: hour
      )

      before_schedule = %Schedule{
        start: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_before = Scheduler.is_schedule_active(before_schedule, ScheduleType.global(), mocked_now)
      assert should_before == false

      during_schedule = %Schedule{
        start: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_during = Scheduler.is_schedule_active(during_schedule, ScheduleType.global(), mocked_now)
      assert should_during == true

      after_schedule = %Schedule{
        start: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_after = Scheduler.is_schedule_active(after_schedule, ScheduleType.global(), mocked_now)
      assert should_after == false
    end)
  end

  test "handles start/end multi day schedules" do

    zero_day = Timex.from_unix(0)

    Enum.each(0..23, fn(hour) ->

      mocked_now = Timex.now()
      |> Timex.set(
        hour: hour,
        minute: 0,
        second: 0,
        microsecond: 0
      )

      zero_day_with_mocked_time = zero_day |> Timex.set(
        hour: hour
      )

      before_schedule = %Schedule{
        start: Timex.add(mocked_now, Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
        end: Timex.add(mocked_now, Timex.Duration.from_days(3)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_before = Scheduler.is_schedule_active(before_schedule, ScheduleType.global(), mocked_now)
      assert should_before == false

      during_schedule = %Schedule{
        start: Timex.add(mocked_now, Timex.Duration.from_days(-1)) |> Timex.to_unix |> Kernel.*(1000),
        end: Timex.add(mocked_now, Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.add(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_during = Scheduler.is_schedule_active(during_schedule, ScheduleType.global(), mocked_now)
      assert should_during == true

      after_schedule = %Schedule{
        start: Timex.add(mocked_now, Timex.Duration.from_days(-3)) |> Timex.to_unix |> Kernel.*(1000),
        end: Timex.add(mocked_now, Timex.Duration.from_days(-1)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.start_end(),
        startTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: Timex.subtract(zero_day_with_mocked_time, Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_after = Scheduler.is_schedule_active(after_schedule, ScheduleType.global(), mocked_now)
      assert should_after == false
    end)
  end


  test "handles daily schedules" do

    zero_day = Timex.from_unix(0)

    Enum.each(0..23, fn(hour) ->

      mocked_now = Timex.now()
      |> Timex.set(
        hour: hour,
        minute: 0,
        second: 0,
        microsecond: 0
      )

      zero_day_with_mocked_time = Timex.set(zero_day,
        hour: mocked_now.hour()
      )

      before_whole_schedule = %Schedule{
        start: mocked_now |> Timex.add(Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
    end: mocked_now |> Timex.add(Timex.Duration.from_days(3)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.daily(),
        startTime: zero_day_with_mocked_time |> Timex.add(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: zero_day_with_mocked_time |> Timex.add(Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_before_whole = Scheduler.is_schedule_active(before_whole_schedule, ScheduleType.global(), mocked_now)
      assert should_before_whole == false, "Hour #{hour}: Should be before whole schedule for schedule with daily time"

      before_schedule = %Schedule{
        start: mocked_now |> Timex.subtract(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.add(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.daily(),
        startTime: zero_day_with_mocked_time |> Timex.add(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: zero_day_with_mocked_time |> Timex.add(Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_before = Scheduler.is_schedule_active(before_schedule, ScheduleType.global(), mocked_now)
      assert should_before == false, "Hour #{hour}: Should be before schedule for schedule with daily time"

      during_schedule = %Schedule{
        start: mocked_now |> Timex.subtract(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.add(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.daily(),
        startTime: zero_day_with_mocked_time |> Timex.subtract(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: zero_day_with_mocked_time |> Timex.add(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_during = Scheduler.is_schedule_active(during_schedule, ScheduleType.global(), mocked_now)
      assert should_during == true, "Hour #{hour}: Should be within schedule for schedule with daily time"

      after_schedule = %Schedule{
        start: mocked_now |> Timex.subtract(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.add(Timex.Duration.from_days(2)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.daily(),
        startTime: zero_day_with_mocked_time |> Timex.subtract(Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: zero_day_with_mocked_time |> Timex.subtract(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_after = Scheduler.is_schedule_active(after_schedule, ScheduleType.global(), mocked_now)
      assert should_after == false, "Hour #{hour}: Should be after schedule for schedule with daily time"

      after_whole_schedule = %Schedule{
        start: mocked_now |> Timex.subtract(Timex.Duration.from_days(3)) |> Timex.to_unix |> Kernel.*(1000),
        end: mocked_now |> Timex.subtract(Timex.Duration.from_days(1)) |> Timex.to_unix |> Kernel.*(1000),
        timezone: "UTC",
        timeType: TimeType.daily(),
        startTime: zero_day_with_mocked_time |> Timex.subtract(Timex.Duration.from_hours(3)) |> Timex.to_unix |> Kernel.*(1000),
        endTime: zero_day_with_mocked_time |> Timex.subtract(Timex.Duration.from_hours(1)) |> Timex.to_unix |> Kernel.*(1000),
      } |> Jason.encode!() |> Jason.decode!()

      should_after_whole = Scheduler.is_schedule_active(after_whole_schedule, ScheduleType.global(), mocked_now)
      assert should_after_whole == false, "Hour #{hour}: Should be after whole schedule for schedule with daily time"
    end)
  end
end
